### PR TITLE
Fcc 18 forgot password

### PR DESCRIPTION
**Pull request by kpennels**
** Updated files **
/Analytics/app.py - the forgot_password.py resource class is imported and the  /forgot_password endpoint is exposed 

**New Files**
/resources/forgot_password.py - Implements the functionality of the /forgot_password POST endpoint. This functionality involves:1) extracting the user's email from the body of the post request 2) create a system generated password 3) send a user an email (using sendgrid) with this system generated password and prompts them to login with this password 4) changing the users password to the system generated password. 

/resources/forgot_password_email.py - Contains the html that will be included in the email sent to the user in the form of a formatted string so that the user's fullname and system generated password can be included 

/resources/test_forgot_password.py - Contains the unit tests for the /forgot_password endpoint

### CODE REVIEW

File: Analytics/resources/forgot_password_email.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
+""" Define the html that is included in a forgot password email. """
```
Comment (j0rdanl0urens) [inline Analytics/resources/forgot_password_email.py]: Is there a reason this is a python file and not html?

File: Analytics/resources/test_forgot_password.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import pytest
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: Did Dan tell you about how you guys must do the unittests?

File: Analytics/resources/test_forgot_password.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    user = Users("James", "james@dotmodus.com", Users.generate_hash(
+        "1234".encode("utf8")).decode("utf8"), True, True)
+
+    try:
+        user.save()
+        user.commit()
+        yield user
+        db.session.delete(user)
+        db.session.commit()
+    except Exception as e:
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: Replace this with the exact exception that is being catched.

File: Analytics/resources/test_forgot_password.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        pass
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: At least log the error

File: Analytics/resources/test_forgot_password.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def test_forgot_password(test_client, dummy_user: db.Model):
+    """
+    Test whether email is sent successfully and password has changed in
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: Add params to docstrings

File: Analytics/resources/test_forgot_password.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        user.save()
+        user.commit()
+        yield user
+        db.session.delete(user)
+        db.session.commit()
+    except Exception as e:
+        pass
+
+
+def test_forgot_password(test_client, dummy_user: db.Model):
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: Type

File: Analytics/resources/test_forgot_password.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    response_forgot = test_client.post('/forgot_password', data=dict(
+        email=current_user.email), follow_redirects=True)
+    response_forgot_json = response_forgot.get_json()
+    assert "success" in response_forgot_json["message"]
+    assert response_forgot.status_code == 200
+
+    updated_user = Users.find_by_email(dummy_user.email)
+    assert forgotten_password != updated_user.password
+
+
```
Comment (j0rdanl0urens) [inline Analytics/resources/test_forgot_password.py]: Remove

File: Analytics/resources/forgot_password.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    def send_forgot_password_email(name: str, email: str,
+                                   new_password: str) -> bool:
+
+        """
+        Send email to a user containing their new system generated password
+        """
+
+        text_version = """\
+            Hi {username}
+
```
[the remark this answers is not in the document]
Comment (kpennels) [reply Analytics/resources/forgot_password.py]: It was supposed to be newline characters

File: Analytics/resources/test_forgot_password.py (hunk 5 above)
Comment (kpennels) [reply Analytics/resources/test_forgot_password.py]: Added params to docstring

**Pull request merged**